pkg/unique-id/module.go

```diff
@@ -0,0 +1,126 @@
+package unique_id
+
+import (
+	"context"
+	"fmt"
+	"sync"
+
+	"github.com/applike/gosoline/pkg/cfg"
+	"github.com/applike/gosoline/pkg/kernel"
+	"github.com/applike/gosoline/pkg/log"
+)
+
+type uniqueIdGenerator struct {
+	kernel.EssentialBackgroundModule
+	kernel.EssentialStage
+
+	logger     log.Logger
+	registerer InstanceRegisterer
+	generator  Generator
+	instance   Instance
+	settings   Settings
+}
+
+var um = struct {
+	sync.Mutex
+	instance *uniqueIdGenerator
+}{}
+
+func ProvideGenerator(config cfg.Config, logger log.Logger) (Generator, error) {
+	if um.instance != nil {
+		return um.instance, nil
+	}
+
+	um.Lock()
+	defer um.Unlock()
+
+	var err error
+	if um.instance, err = NewGenerator(config, logger); err != nil {
+		return nil, err
+	}
+
+	return um.instance, nil
+}
+
+func ProvideModule() kernel.ModuleFactory {
+	return func(_ context.Context, config cfg.Config, logger log.Logger) (kernel.Module, error) {
+		if um.instance != nil {
+			return um.instance, nil
+		}
+
+		um.Lock()
+		defer um.Unlock()
+
+		var err error
+		if um.instance, err = NewGenerator(config, logger); err != nil {
+			return nil, err
+		}
+
+		return um.instance, nil
+	}
+}
+
+func NewGenerator(config cfg.Config, logger log.Logger) (*uniqueIdGenerator, error) {
+	settings := Settings{}
+	config.UnmarshalKey("unique_id", &settings)
+
+	logger = logger.WithChannel("unique-id")
+
+	registerer, err := NewInstanceRegisterer(config, logger, settings)
+	if err != nil {
+		return nil, fmt.Errorf("could not create instance registerer: %w", err)
+	}
+
+	// pass application context once available
+	instance, err := registerer.Register(context.Background())
+	if err != nil {
+		return nil, fmt.Errorf("could not register instance: %w", err)
+	}
+
+	logger = logger.WithFields(log.Fields{
+		"machine_id": instance.MachineId,
+	})
+
+	logger.Info("registered instance")
+
+	generator := generatorProvider(instance.MachineId, settings.StartTime)
+	if generator == nil {
+		return nil, fmt.Errorf("could not create sonyflake for instance %d", instance.MachineId)
+	}
+
+	return NewUniqueIdWithInterfaces(logger, registerer, generator, *instance), nil
+}
+
+func NewUniqueIdWithInterfaces(logger log.Logger, registerer InstanceRegisterer, generator Generator, instance Instance) *uniqueIdGenerator {
+	return &uniqueIdGenerator{
+		logger:     logger,
+		registerer: registerer,
+		generator:  generator,
+		instance:   instance,
+	}
+}
+
+func (m *uniqueIdGenerator) Run(ctx context.Context) error {
+	<-ctx.Done()
+
+	// prevent further generations
+	m.generator = nil
+
+	// need a new context here, because the run context is already finished
```

Comment:
Hmm, should your context eventually cancel after a timeout? Otherwise you could block here if something needs a lot of retries... and the kernel would kill itself. However, not sure if that really makes a difference

Reply:
Always depends on the shutdown scenario. I don't think that will really happen, but I added a 5 second timeout to the context